File: Algorithms/src/3228. Maximum Number of Operations to Move Ones to the End/Code/solution.py
class Solution:
    def maxOperations(self, s: str) -> int:
        # ans and cnt use Python int (arbitrary precision)
        ans = 0
        cnt = 0  # number of '1's seen so far
        n = len(s)
        for i, c in enumerate(s):
            if c == '1':
                cnt += 1
            else:  # c == '0'
                # If this zero is immediately after a '1', it will be "used"
                # by each previously seen '1' across the optimal sequence.
                if i > 0 and s[i-1] == '1':
                    ans += cnt
        return ans

File: Algorithms/src/3228. Maximum Number of Operations to Move Ones to the End/Code/solution.cpp
class Solution
{
public:
    int maxOperations(string s)
    {
        long long ans = 0;  // result can be up to O(n^2) in value -> use 64-bit
        long long ones = 0; // count of '1's seen so far
        int n = s.size();
        for (int i = 0; i < n; ++i)
        {
            if (s[i] == '1')
            {
                ++ones; // saw another '1'
            }
            else
            { // s[i] == '0'
                // if this zero sits immediately after a '1', then all previous '1's
                // contribute one operation for this zero over the whole process.
                if (i > 0 && s[i - 1] == '1')
                    ans += ones;
            }
        }
        return (int)ans;
    }
};
